### PR TITLE
ShapeInfo<...>::reinit: add missing instantiations

### LINKED ISSUE #4722: Linker errors in tests on MacOS
Failing tests:
- lac/constraints_local_to_global_rect
```
1894: Undefined symbols for architecture x86_64:
1894:   "void dealii::ConstraintMatrix::distribute_local_to_global<dealii::FullMatrix<double> >(dealii::FullMatrix<dealii::FullMatrix<double>::value_type> const&, std::__1::vector<unsigned int, std::__1::allocator<unsigned int> > const&, dealii::ConstraintMatrix const&, std::__1::vector<unsigned int, std::__1::allocator<unsigned int> > const&, dealii::FullMatrix<double>&) const", referenced from:
1894:       test() in constraints_local_to_global_rect.cc.o
```
- matrix_free/assemble_matrix_01
```
2394: Undefined symbols for architecture x86_64:
2394:   "void dealii::internal::MatrixFreeFunctions::ShapeInfo<dealii::VectorizedArray<double> >::reinit<2>(dealii::Quadrature<1> const&, dealii::FiniteElement<2, 2> const&, unsigned int)", referenced from:
2394:       dealii::internal::MatrixFreeFunctions::ShapeInfo<dealii::VectorizedArray<double> >::ShapeInfo<2>(dealii::Quadrature<1> const&, dealii::FiniteElement<2, 2> const&, unsigned int) in assemble_matrix_01.cc.o
2394:   "void dealii::internal::MatrixFreeFunctions::ShapeInfo<dealii::VectorizedArray<double> >::reinit<3>(dealii::Quadrature<1> const&, dealii::FiniteElement<3, 3> const&, unsigned int)", referenced from:
2394:       dealii::internal::MatrixFreeFunctions::ShapeInfo<dealii::VectorizedArray<double> >::ShapeInfo<3>(dealii::Quadrature<1> const&, dealii::FiniteElement<3, 3> const&, unsigned int) in assemble_matrix_01.cc.o
```
- matrix_free/assemble_matrix_02: Same as above
- matrix_free/assemble_matrix_03 : Same as above
- matrix_free/matrix_vector_15: Same as above
- matrix_free/matrix_vector_16: Same as above

### PR DESCRIPTION
**Pull request by tamiko**
In reference to #4722
In reference to #4704

### CODE REVIEW

File: doc/news/changes/minor/20170809ArndtMaierPelteret [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Fixed: Debugged and fixed a long standing linker error due to missing
+instantiations of ShapeInfo::reinit variants.
```
Comment (bangerth) [inline doc/news/changes/minor/20170809ArndtMaierPelteret]: Is this only on some platforms? That would be worth mentioning.

File: source/matrix_free/matrix_free.inst.in [hunk 2]
```diff
@@ -39,6 +39,12 @@ for (deal_II_dimension : DIMENSIONS)
     template void internal::MatrixFreeFunctions::ShapeInfo<float>::reinit
     <deal_II_dimension>(const Quadrature<1> &, const FiniteElement
                         <deal_II_dimension,deal_II_dimension> &, const unsigned int);
+    template void internal::MatrixFreeFunctions::ShapeInfo<VectorizedArray<double> >::reinit
+    <deal_II_dimension>(const Quadrature<1> &, const FiniteElement
```
Comment (bangerth) [inline source/matrix_free/matrix_free.inst.in]: do me a favor here and below, and keep the template argument at the function name. That is,
```
template 
void 
internal::MatrixFreeFunctions::ShapeInfo<VectorizedArray<double> >::
reinit<deal_II_dimension>(const Quadrature<1> &, const FiniteElement
```
or similar.

File: doc/news/changes/minor/20170809ArndtMaierPelteret (hunk 1 above)
Comment (tamiko) [inline doc/news/changes/minor/20170809ArndtMaierPelteret]: @bangerth It is fairly generic - only depends on whether the optimizer inlines reinit<> in another compilation unit (we reproduced on mac and linux with clang, and two gcc versions).

**Comment by bangerth** [comment]
OK with the two minor edits.

File: doc/news/changes/minor/20170809ArndtMaierPelteret (hunk 1 above)
Comment (kronbichler) [inline doc/news/changes/minor/20170809ArndtMaierPelteret]: It seems that you only add instantiations for reinit of `ShapeInfo<VectorizedArray<double>>` and `ShapeInfo<VectorizedArray<float>>`. I merely added this code a week ago in #4680. So I wouldn't really call this a long-standing linker error... ;-)

**Comment by kronbichler** [comment]
But thanks for fixing up this oversight of mine.

File: doc/news/changes/minor/20170809ArndtMaierPelteret (hunk 1 above)
Comment (tamiko) [inline doc/news/changes/minor/20170809ArndtMaierPelteret]: @kronbichler Shall we change that? :-)

Comment (kronbichler) [inline doc/news/changes/minor/20170809ArndtMaierPelteret]: I guess we should remove the 'long standing', yes. Can you do it? (I'm away from my computer now...)

**Pull request merged**